Epic 6800 : Issue Definition designer (Story 3)- Issues module implementation

labkey-bpatel opened the pull request:
#### Rationale
This PR includes setup and new development to convert Issues Definition designer to a react-based designer, available via enabling Experimental feature flag.

#### Related Pull Requests
https://github.com/LabKey/labkey-ui-components/pull/245

#### Changes
- Includes setup of Issues module for react page development
- Includes entrypoint for Issue’s Designer usage
- Includes URL param parsing, loading the existing issue model for edit and navigation on success
- Experimental feature flag
- New issues api actions to getProjectGroups and getUsersForGroup

labkey-susanh committed:
First rename to experiment2

labkey-susanh committed:
Then rename to experiment

labkey-susanh committed:
Update references with renamed directory

labkey-susanh committed:
Update references with renamed directory

labkey-bpatel committed:
Add experimental feature to use React-based Issues Designer.

labkey-bpatel committed:
Setup and entry point for react dev.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_renameTwice' into fb_issuesdesigner_story3_6800

labkey-bpatel committed:
Updates to render Fields.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_issuesdesigner_story3_6800

labkey-bpatel committed:
Handle navigation on Save.

labkey-bpatel committed:
Handle permission

labkey-bpatel committed:
Add customEmailTemplateButton to Issues List grid (in the new designer this button will not be present unlike GWT designer, hence moving it to the grid).

labkey-bpatel committed:
Add issues.getUsersForGroup & pojos

labkey-bpatel committed:
Add action getProjectGroups from Issues controller. Update package.json with alpha version.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_issuesdesigner_story3_6800

cnathe committed:
Merge branch 'develop' into fb_issuesdesigner_story3_6800

cnathe commented:
@labkey-bpatel Don't forget about items `1.iv` and `4` in the "Setting up a LabKey module doc here: https://github.com/LabKey/platform/blob/develop/webpack/README.md

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1968,21 +1977,32 @@ public ModelAndView getView(IssuesController.AdminForm adminForm, BindException
-            Domain domain = issueListDef.getDomain(getUser());
+            else
+            {
+                if (issueListDef == null)
+                {
+                    return new HtmlView(getUndefinedIssueListMessage(getViewContext(), issueDefName));
+                }
```
This check for `issueListDef == null` and returning the HtmlView should likely stay out of this if/else block. It will still apply for the new designer ui as it is a way to show an error message if the URL param is invalid.

labkey-bpatel replied:
ah youre right, now fixed

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1997,6 +2017,205 @@ public NavTree appendNavTrail(NavTree root)
+            List<Principal> groups = new ArrayList<>();
+
+            SecurityManager.getGroups(getContainer().getProject(), true).stream().filter(group -> !group.isGuests() && (!group.isUsers() || getUser().hasRootAdminPermission())).forEach(group -> {
+                String displayText = (group.isProjectGroup() ? "" : "Site:") + group.getName();
```
I think you previously had a space added after the `Site:` string. It would be nice to put that back in.

labkey-bpatel replied:
yes! done

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1997,6 +2017,205 @@ public NavTree appendNavTrail(NavTree root)
+                String displayText = (group.isProjectGroup() ? "" : "Site:") + group.getName();
+
+                Principal principal = new Principal();
+                principal.setActive(group.isActive());
+                principal.setDisplayName(displayText);
+                principal.setName(group.getName());
+                principal.setType(group.getType());
+                principal.setUserId(group.getUserId());
```
Looking at the client side code, I think we only end up using the group id and the displayName. Do you think we need to return the rest of this information?

cnathe replied:
Alternatively, could we just build up a `List<Group>`'s here and then return that from the API? then I think we could do the displayText part on the client side since the Group object would include the `isProjectGroup` info. That way we could do away with this extra `public static class Principal` class.

labkey-bpatel replied:
Nope, not needed, I've simplified it now, let me know what you think!

cnathe replied:
Looks good

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1997,6 +2017,205 @@ public NavTree appendNavTrail(NavTree root)
+
+    }
+
+    public static class Principal
```
these class names seem backwards, isn't this one for a "Group"?

labkey-bpatel replied:
This class was same as client side Principle model. But its really an overkill, so now removed.

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1997,6 +2017,205 @@ public NavTree appendNavTrail(NavTree root)
+            this.active = active;
+        }
+    }
+    public static class UserGroup
```
these class names seem backwards, isn't this one for a "User/Principal"?

labkey-bpatel replied:
This class was the same as client side's UserGroup - but now removed.

cnathe commented on issues/src/org/labkey/issue/IssuesController.java:
```diff
@@ -1997,6 +2017,205 @@ public NavTree appendNavTrail(NavTree root)
+                        if (getContainer().hasPermission(user, UpdatePermission.class))
+                        {
+                            UserGroup usergrp = new UserGroup();
+                            usergrp.setGroupId(group.getUserId());
+                            usergrp.setGroupName(group.getName());
+                            usergrp.setUserId(user.getUserId());
+                            usergrp.setUserName(user.getDisplayName(getUser()));
```
Similar question as the other response object. Does the client side code use/need all of this? I assume it only cares about the id and displayName.

labkey-bpatel replied:
nope not needed, please review my updates.

cnathe replied:
Looks good

cnathe commented on issues/src/client/Designer/Designer.tsx:
```diff
@@ -0,0 +1,149 @@
+    isLoadingModel: boolean,
+    message?: string,
+    model?: IssuesListDefModel,
+    hasDesignIssuesAdminPermission?: boolean
```
I don't think we need this as I don't think there is a specific permissions class for issue list def admins. I believe it is just a container admin check which we already have on the IssuesController.AdminAction.

labkey-bpatel replied:
ah, good point, now removed

cnathe commented on issues/src/client/Designer/Designer.tsx:
```diff
@@ -0,0 +1,149 @@
+    componentDidMount() {
+        const issueDefName = ActionURL.getParameter('issueDefName');
+
+        Security.getUserPermissions({
+            containerPath: getServerContext().container.path,
+            success: (data) => {
+                this.setState(() => ({
+                    hasDesignIssuesAdminPermission: data.container.effectivePermissions.indexOf(PermissionTypes.Admin) > -1
+                }));
+            },
+            failure: (error) => {
+                this.setState(() => ({
+                    message: error.exception,
+                    hasDesignIssuesAdminPermission: false
+                }));
+            }
+        });
```
I think this can all be removed as the action requires the permissions and the entryPoint requires admin as well.

labkey-bpatel replied:
done

cnathe commented on issues/src/client/Designer/Designer.tsx:
```diff
@@ -0,0 +1,149 @@
+            return <LoadingSpinner/>
+        }
+
+        if (!hasDesignIssuesAdminPermission) {
+            return <Alert>You do not have sufficient permissions to create or edit a issues list definition design.</Alert>
+        }
```
I think this can be removed as well.

labkey-bpatel replied:
done

labkey-bpatel committed:
Code Review change:
- Move issueListDef null check outside of if-else
- Simplify and combine java beans.
- Remove setting of unneeded values.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/fb_issuesdesigner_story3_6800' into fb_issuesdesigner_story3_6800

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_issuesdesigner_story3_6800

labkey-bpatel committed:
Update labkey/components version.

labkey-bpatel committed:
Code Review change: Remove permission from the state, not needed.

cnathe reviewed (changes requested):
Code changes look good. I think the only thing left are those two items missing from the "Setting up a LabKey module" doc for the README file and .gitignore changes: https://github.com/LabKey/platform/blob/develop/webpack/README.md

cnathe committed:
Issues designer fix for navigation on cancel button click

labkey-bpatel committed:
Add README and update .gitignore.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_issuesdesigner_story3_6800

labkey-bpatel committed:
Update labkey/components version.
